src/runtime-rs/crates/hypervisor/src/device/driver/port_device.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2024 Ant Group
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+//use std::collections::HashMap;
+
+use anyhow::{anyhow, Context, Result};
+use async_trait::async_trait;
+
+use crate::device::{
+    hypervisor,
+    topology::{PCIePort, PCIePortBusPrefix, PCIeTopology, TopologyPortDevice, DEFAULT_BRIDGE_BUS},
+    Device, DeviceType,
+};
+
+#[derive(Debug, Default, Clone)]
+pub struct PortDeviceConfig {
+    pub port_type: PCIePort,
+    pub total_ports: u32,
+    pub memsz_reserve: u64,
+    pub pref64_reserve: u64,
+}
+
+impl PortDeviceConfig {
+    pub fn new(port_type: PCIePort, total_ports: u32) -> Self {
+        Self {
+            port_type,
+            total_ports,
+            // FIXME:
+            // A method to automatically determine the maximum memory size
+            // based on all vfio devices' information on the host is coming soon.
+            memsz_reserve: 2097152_u64,
+            pref64_reserve: 4194304_u64,
+        }
+    }
+}
+
+#[derive(Debug, Default, Clone)]
+pub struct PCIePortDevice {
+    /// device id for sharefs device in device manager
+    pub device_id: String,
+
+    /// config for sharefs device
+    pub config: PortDeviceConfig,
+}
+
+impl PCIePortDevice {
+    pub fn new(device_id: &str, config: &PortDeviceConfig) -> Self {
+        Self {
+            device_id: device_id.to_string(),
+            config: config.clone(),
+        }
+    }
+}
+
+#[async_trait]
+impl Device for PCIePortDevice {
+    async fn attach(
+        &mut self,
+        pcie_topo: &mut Option<&mut PCIeTopology>,
+        h: &dyn hypervisor,
+    ) -> Result<()> {
+        if let Some(topology) = pcie_topo {
+            if let Some(devices) = topology
+                .pcie_devices_per_port
+                .get_mut(&self.config.port_type)
+            {
+                match self.config.port_type {
+                    PCIePort::RootPort => {
+                        for index in 0..topology.pcie_root_ports {
+                            let rp_id = format!("{}{}", PCIePortBusPrefix::RootPort, index);
+                            devices.push(TopologyPortDevice::new(
+                                &rp_id,
+                                DEFAULT_BRIDGE_BUS,
```

Comment:
Is this intentional, or should a root port belong to `DEFAULT_PCIE_ROOT_BUS`?

There are several more places in this PR that seem to associate root port devices with `DEFAULT_BRIDGE_BUS` so I'd like to clear this up.